docs/guidance/prometheus-grafana.md

```diff
@@ -0,0 +1,191 @@
+
+## Step 6: Collect Worker Node metrics with PodMonitors
+
+KubeRay operator does not create a Kubernetes service for the Ray worker Pods, therefore we cannot use a Prometheus ServiceMonitor to scrape the metrics from the worker Pods. To collect worker metrics, we can use `Prometheus PodMonitors CRD` instead.
+
+**Note**: We could create a Kubernetes service with selectors a common label subset from our worker pods, however, this is not ideal because our workers are independent from each other, that is, they are not a collection of replicas spawned by replicaset controller. Due to that, we should avoid using a Kubernetes service for grouping them together.
+
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: PodMonitor
+metadata:
+  name: ray-workers-monitor
+  namespace: prometheus-system
+  labels:
+    # `release: $HELM_RELEASE`: Prometheus can only detect PodMonitor with this label.
+    release: prometheus
+    ray.io/cluster: raycluster-kuberay # $RAY_CLUSTER_NAME: "kubectl get rayclusters.ray.io"
+spec:
+  jobLabel: ray-workers
+  # Only select Kubernetes Pods in the "default" namespace.
+  namespaceSelector:
+    matchNames:
+      - default
+  # Only select Kubernetes Pods with "matchLabels".
+  selector:
+    matchLabels:
+      ray.io/node-type: worker
+  # A list of endpoints allowed as part of this PodMonitor.
+  podMetricsEndpoints:
+  - port: metrics
+```
+* **PodMonitor** in `namespaceSelector` and `selector` are used to select Kubernetes Pods.
+  ```sh
+  kubectl get pod -n default -l ray.io/node-type=worker
+  # NAME                                          READY   STATUS    RESTARTS   AGE
+  # raycluster-kuberay-worker-workergroup-5stpm   1/1     Running   0          3h16m
+  ```
+
+* `ray.io/cluster: $RAY_CLUSTER_NAME`: We also define `metadata.labels` by manually adding `ray.io/cluster: <ray-cluster-name>` and then instructing the PodMonitors resource to add that label in the scraped metrics via `spec.podTargetLabels[0].ray.io/cluster`.
+
+## Step 7: Access Prometheus Web UI
+```sh
+# Forward the port of Prometheus Web UI in the Prometheus server Pod.
+kubectl port-forward --address 0.0.0.0 prometheus-prometheus-kube-prometheus-prometheus-0 -n prometheus-system 9090:9090
+
+# Check ${YOUR_IP}:9090/targets for the Web UI (e.g. 127.0.0.1:9090/targets)
+# You should be able to see "podMonitor/prometheus-system/ray-workers-monitor/0 (1/1 up)"
+# and "serviceMonitor/prometheus-system/ray-head-monitor/0 (1/1 up)" in the page.
+```
+
+![Prometheus Web UI](../images/prometheus_web_ui.png)
+
+## Step 8: Access Grafana
+
+```sh
+# Forward the port of Grafana
+kubectl port-forward --address 0.0.0.0 deployment/prometheus-grafana -n prometheus-system 3000:3000
```

Comment:
Got this error at this step:
```
Unable to listen on port 3000: Listeners failed to create with the following errors: [unable to create listener: Error listen tcp4 0.0.0.0:3000: bind: address already in use]
error: unable to listen on any of the requested ports: [{3000 3000}]
```

But I'm not sure what could be using port 3000.  Do I need to cancel one of the previous steps?

Reply:
Looks like I already have grafana running for some reason, I think it's unrelated to this tutorial: 
```
sudo lsof -i :3000
Password:
COMMAND    PID   USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
grafana-s 3915 archit   21u  IPv6 0xd6d807306e5fa78f      0t0  TCP *:hbci (LISTEN)
```

Let me figure out how to stop it. I don't know if it's a common enough situation to mention in this document.

Reply:
Naively killed the process but it kept restarting. The solution here https://unix.stackexchange.com/questions/601063/port-3000-is-always-being-hogged-by-grafana-server
is to run `brew services stop grafana`.  Probably don't need to mention in this doc